core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedService.java

```diff
@@ -251,7 +251,30 @@ Route route() {
 
     @Override
     public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) throws Exception {
-        final HttpResponse response = HttpResponse.from(serve0(ctx, req));
+        if (!defaultHttpHeaders.isEmpty()) {
+            ctx.mutateAdditionalResponseHeaders(mutator -> mutator.add(defaultHttpHeaders));
+        }
+        if (!defaultHttpTrailers.isEmpty()) {
+            ctx.mutateAdditionalResponseTrailers(mutator -> mutator.add(defaultHttpTrailers));
+        }
+
+        final AggregationType aggregationType =
+                AnnotatedValueResolver.aggregationType(aggregationStrategy, req.headers());
+
+        final HttpResponse response;
+        if (aggregationType == AggregationType.NONE && !useBlockingTaskExecutor) {
+            // Fast-path: No aggregation required and blocking task executor is not used.
+            final Object result = invoke(ctx, req, AggregatedResult.EMPTY);
+            if (responseType == ResponseType.HTTP_RESPONSE) {
+                assert result != null;
+                response = (HttpResponse) result;
+            } else {
+                response = convertResponse(ctx, result);
```

Comment:
Can we just call
```suggestion
            response = convertResponse(ctx, invoke(ctx, req, AggregatedResult.EMPTY));
```
because `convertResponse` returns right away if the `result` is an instanceof `HttpResponse`?

Reply:
Good point! Updated :+1:

Reply:
[this comment](https://github.com/line/armeria/pull/4888#discussion_r1227657232) made me change the code a little bit more. At second thought, it's probably fine not calling `convertResponse` to save the potentially high cost of `instanceof`.

Reply:
Got it. Thanks!